official/projects/lra/linformer_encoder.py

```diff
@@ -0,0 +1,294 @@
+
+    self._embedding_norm_layer = tf.keras.layers.LayerNormalization(
+        name='embeddings/layer_norm', axis=-1, epsilon=1e-12, dtype=tf.float32)
+
+    self._embedding_dropout = tf.keras.layers.Dropout(rate=output_dropout,
+                                                      name='embedding_dropout')
+
+    # We project the 'embedding' output to 'hidden_size' if it is not already
+    # 'hidden_size'.
+    self._embedding_projection = None
+    if embedding_width != hidden_size:
+      self._embedding_projection = tf.keras.layers.EinsumDense(
+          '...x,xy->...y',
+          output_shape=hidden_size,
+          bias_axes='y',
+          kernel_initializer=initializer,
+          name='embedding_projection')
+
+    self._transformer_layers = []
+    self._attention_mask_layer = layers.SelfAttentionMask(
+        name='self_attention_mask')
+    for i in range(num_layers):
+      layer = LinformerEncoderBlock(
+          num_attention_heads=num_attention_heads,
+          low_rank_features=low_rank_features,
+          inner_dim=inner_dim,
+          inner_activation=inner_activation,
+          output_dropout=output_dropout,
+          attention_dropout=attention_dropout,
+          norm_first=norm_first,
+          return_attention_scores=False,
+          kernel_initializer=tf_utils.clone_initializer(initializer),
+          name='transformer/layer_%d' % i)
+      self._transformer_layers.append(layer)
+    self._num_layers = num_layers
+
+    self._pooler_layer = tf.keras.layers.Dense(units=hidden_size,
+                                               activation='tanh',
+                                               kernel_initializer=initializer,
+                                               name='pooler_transform')
+
+    self._config = {
+        'vocab_size': vocab_size,
+        'hidden_size': hidden_size,
+        'num_layers': num_layers,
+        'low_rank_features': low_rank_features,
+        'num_attention_heads': num_attention_heads,
+        'max_sequence_length': max_sequence_length,
+        'type_vocab_size': type_vocab_size,
+        'inner_dim': inner_dim,
+        'inner_activation': tf.keras.activations.serialize(activation),
+        'output_dropout': output_dropout,
+        'attention_dropout': attention_dropout,
+        'initializer': tf.keras.initializers.serialize(initializer),
+        'output_range': output_range,
+        'embedding_width': embedding_width,
+        'embedding_layer': embedding_layer,
+        'norm_first': norm_first
+    }
+    self.inputs = dict(input_word_ids=tf.keras.Input(shape=(None,),
+                                                     dtype=tf.int32),
+                       input_mask=tf.keras.Input(shape=(None,), dtype=tf.int32),
+                       input_type_ids=tf.keras.Input(shape=(None,),
+                                                     dtype=tf.int32))
+
+  def call(self, inputs):
+    word_embeddings = None
+
+    if isinstance(inputs, dict):
```

Comment:
Can we refactor this out?

Reply:
refactored! let me know if you meant something else